### DISCUSSION TITLE
GST Prosody transfer on Tacotron 2 is not working

### DISCUSSION THREAD

**Discussion by saibharani** [coqui-ai/TTS#1723] [General Q&A]
### Describe the bug

I trained an Tacotron2 GST model on LJspeech dataset and own Emotional dataset for 100k steps using `use_gst=True,
    gst=GSTConfig(),` options in training.

### To Reproduce

But during inference the audio sounds same with or without using style_wav. I used `from TTS.utils.synthesizer import Synthesizer` function for synthesizing the text. Can you suggest any changes in config or any corrections to improve the prosody transfer quality.

### Environment

```shell
- TTS version 0.7.0
- Pytorch version 1.9.0
- CUDA version 10.2
- OS Ubuntu 18.04
- Python 3.8.13
```

### Additional context

Any help is appreciated, thanks.

**Comment by WeberJulian** [comment]
Hey, LJSpeech might not be the best dataset to train GST on because of the lack of prosody. In the paper they use data from the 2013 Blizzard Challenge. But you might try with audio book data as well since it tends to be more expressive.

**Comment by WeberJulian** [comment]
You can also try inference by tweaking manually the style tokens to see if you can get variations.

**Comment by saibharani** [comment]
That makes sense I will try to train a new model using Blizzard dataset or any Audio book data and will give an update here.

Can you please elaborate on tweaking style tokens manually, do you mean by inputting a dict to gst_style_input_weights parameter? can you give any example of manual style tokens?

Thanks

**Reply by erogol** [reply]
Yes, you can try different weights and see how the output changes. There is no recipe for the right weights.

**Comment by erogol** [comment]
I move it as it is not a functional issue.

**Comment by WeberJulian** [comment]
Hey @saibharani, did you get better results with another dataset?

**Reply by saibharani** [reply]
I tried on the VCTK dataset for about 75K steps, but I didn't see any good improvement in the GST Quality I am thinking of trying dev-emotion or vits-pitch-pred branch and see if I can get any good prosody transfer. Are there any plans to release models for any of those recipies?

**Reply by WeberJulian** [reply]
VCTK is even more flat in prosody than LJSpeech. If you don't use Blizard, try something from audio books, like a single speaker from MAILABS.

> I am thinking of trying dev-emotion or vits-pitch-pred

I don't think they'll work well out of the box, it's still in research phase

> Are there any plans to release models for any of those recipies?

I don't think so

**Reply by bharaniyv** [reply]
I currently don't have Blizard or MAILABS datasets with me but I am planning on training with MAILABS for the multi lingual support. I will report the results when I am done with training and testing it.

> I don't think they'll work well out of the box, it's still in research phase

How about the FastSpeech model I think FastSpeech also has duration, and Pitch prediction which partially works as a prosody transfer. Do you think they will work? Have you tried training any of those model and how is their quality?

**Reply by WeberJulian** [reply]
Here is the download link https://www.caito.de/2019/01/03/the-m-ailabs-speech-dataset/
Fastpitch looses a bit in quality compared to VITS, but yeah you can give it a go.
The best model I tried for prosody transfer is Capacitron.

**Reply by saibharani** [reply]
Thanks for the download link, I also tried VITS but the prosody in VITS is not good now, maybe it will be better when the vits-pitch-pred model gets completed.

I tried capacitron too but it requires the ground truth sample and text for inference but I am interested in a model that doesn't require ground truth text. The model I trained on Capacitron using VCTK, didn't give any better results after 50k+ steps compared to GST so I left that model.

**Reply by WeberJulian** [reply]
No it doesn't require, you can sample from prior.
Yeah Capacitron is untested for multispeaker.

**Reply by saibharani** [reply]
Which model do you think will work better for Prosody or emotion transfer among FastSpeech, FastPitch, and SpeedySpeech models? I can see that the FastSpeech2 model is still under development. I want to train any of the above models on MAILABS dataset can you suggest the best option.

**Reply by WeberJulian** [reply]
There is no explicit prosody transfer for the models you cited. But you can transfer prosody from one audio if the text matches exactly by using the same pitch/energy/duration values.

**Reply by saibharani** [reply]
And can these ForwardTTS models be made MultiLingual? If yes can you point me to any example implementations.

**Reply by WeberJulian** [reply]
I don't think it has been tested yet, but it shouldn't be too hard to do.

**Reply by saibharani** [reply]
I was referring to the LanguageEmbedding layer in VITS can it be implemented in ForwardTTS?

**Reply by saibharani** [reply]
and can we add new languages once the LanguageEmbedding layer is trained in VITS? I have a model in VITS and I want to add new Languages to it

**Reply by WeberJulian** [reply]
Yeah but you have to do it manually.

**Comment by jaggzh** [comment]
Can anyone post example code for tacotron2 + gst training and inference?  (Possibly with the style tokens provided during inference, but I can probably get that in there myself.)
